---

ttitle: "`r glue::glue('{params$author_acronym}i{params$investigation}_s{params$study}_a{params$assay}')`"
author:
  - name: "Marilene P.G.P. Schellekens "
    url: "https://orcid.org/0000-0001-9549-1719"
    affiliation: "Leiden University"
    affiliation_url: "https://ror.org/027bh9e22"
date: "`r Sys.Date()`"
output:
  html_document:
    code_download: true
    code_folding: show
    theme: united
    highlight: tango
    df_print: paged
    toc: true
    number_sections: true
    toc_float:
      collapsed: false
      smooth_scroll: true
params:
  author_acronym: "MPGPS"
  project: "MCF7_1"
  investigation: 1
  study: 1
  assay: 1
  system: "MCF7"
---

# ISA {.tabset .tabset-fade .tabset-pills}

## investigation

> RISKHUNT3R WP5 CS5: TXG-MAPr platform for quantitative hazard assessment

Toxicogenomics is a mature method to understand mechanisms of toxicity of a chemical and to quantitatively inform on its hazard, however interpretative analysis tools are suboptimal for risk assessment. We plan to develop and expand a suite of tools, the TXG-MAPr, where gene co-regulated network analysis is applied to large toxicogenomics datasets from relevant test systems covering the target organs and endpoints of interest of the project, in particular for the liver, kidney, neuronal systems, lungs and for non-genotoxic carcinogenicity. With the TXG-MAPr one can: 1) determine the mode of action of a compound, 2) perform compound comparison with legacy data (read-across) and 3) determine mechanistically interpretable, translational biomarkers.

```{r investigation}
params$investigation
```

## study

> Establish the human iPSC-derived HLC TXG-MAPr

In this study, we used a large data set generated by the USEPA of human relevant transctiptional responses. MCF7 cells were exposed to cytokines, growth factors, reference chemicals and ER at 6 hours in 3 biological replicates. In this project the total number of unique treatment chemicals is 1751 and the total numer of samples is 46063


```{r study}
params$study
```

## assay

> TempO-Seq targeted RNA-Sequencing

In the analysis below, we show our QC on raw count data obtained from BioClavis on project MCF7 on samples and sequencing probes and perform differential gene expression analysis. The human Whole Transcriptome v2.1 panel with standard attenuators was used. Replicate correlation was based on log2CPM data. For the PCA plots the qc flags from the USEPA were also included. 

```{r assay}
params$assay
```


# Setup {.tabset .tabset-fade .tabset-pills}

## packages

Package repositories [CRAN](https://cran.r-project.org/) or [Bioconductor](https://www.bioconductor.org/).

```{r Setup-packages}
# CRAN
install.packages(
  pkgs = 
    setdiff(
      x = c(
        "janitor", 
        "devtools", 
        "data.table", 
        "tidyverse", 
        "BiocManager",
        "glue",
        "ggfortify",
        "patchwork",
        "RColorBrewer",
        "ashr"
      ), 
      y = rownames(
        x = installed.packages()
      )
    ),
  repos = "https://cloud.r-project.org/",
  quiet = TRUE
)

# BiocManager
suppressMessages(
  BiocManager::install(
    pkgs = 
      setdiff(
        x = c(
          "DESeq2",
          "ComplexHeatmap"
        ), 
        y = rownames(
          x = installed.packages()
        )
      ),
    update = FALSE, 
    ask = TRUE
  )
)
```

## knitr

Set global chuck options. Global chuck options can be overwritten in individual chunk headers.

```{r Setup-knitr}
knitr::opts_chunk$set(
  echo = TRUE,
  include = TRUE,
  message = FALSE,
  warning = FALSE,
  out.width = "200%",
  tidy.opts = list(width.cutoff = 80)
)
```

## library

Load packages. To resolve function conflicts see [conflicted](https://conflicted.r-lib.org/).

```{r Setup-library}
library(janitor)
library(devtools)
library(data.table)
library(tidyverse)
library(readxl) # from tidyverse
library(glue)
library(ggfortify)
library(patchwork)
library(DESeq2)
library(ComplexHeatmap)
library(RColorBrewer)
library(ashr)
```

## directory

```{r Setup-directory}
# set inputPath
inputPath <- getwd()

# set outputPath
outputPath <- file.path(getwd(), 
                        "OUTPUT")

# create outputPath
if(!exists(outputPath)){
  dir.create(
    path = outputPath, 
    showWarnings = FALSE
  )
}

# set outputFigures
outputFigures <- file.path(getwd(), 
                        "MCF7_TempOSeq_output",
                        "Figures")

# create outputPath
if(!exists(outputFigures)){
  dir.create(
    path = outputFigures, 
    showWarnings = FALSE
  )
}

# outputPathDESeq 
outputPathDESeq <- file.path(getwd(), 
                             "MCF7_TempOSeq_output",
                             "DEseq")

# create outputPathDESeq
if(!exists(outputPathDESeq)){
  dir.create(
    path = outputPathDESeq, 
    showWarnings = FALSE
  )
}
```

## session_info

Save R session info as JSON file and print.

```{r Setup-session_info}
# write session_info to json
devtools::session_info() |>
  jsonlite::write_json(path = file.path(outputPath, 
                                        "session_info.json"))

# print session_info
devtools::session_info()
```

## function {.tabset .tabset-fade .tabset-pills}

### tidy_check

```{r tidy_check}
tidy_check = function(countdata, metadata) {
  # We assume first column will contain the row names (tidy = T in DESeqDataSetFromMatrix)
  names_count = names(countdata)[-1]
  # Not obligatory for DEseq2, but we want row names in the first column
  names_meta = metadata[[1]]
  # Test if the names from the count and metadata are identical (same entries, same order)
  identical = identical(names_count, names_meta)
  # If not identical, do we have the same entries (ignores duplicates)
  setequal = identical || setequal(names_count, names_meta)
  # Test for duplicates (in theory, but very unlikely, this can also happen if identical is true)
  has_duplicates = anyDuplicated(names_meta) > 0
  # If the names are not identical, but we're looking at the same entries without duplicates,
  # then the problem is the order of the entries
  problem_is_order = !identical && setequal && !has_duplicates
  # List possibilities to test
  input = list(identical = identical, setequal = setequal, has_duplicates = has_duplicates, problem_is_order = problem_is_order)
  
  if(input$identical & input$setequal & !input$has_duplicates & !input$problem_is_order){
    return(TRUE)
  } else {
    print(input)
    return(FALSE)
  }
}
```

### runDESeq

```{r runDESeq}
runDESeq <- function(countdata, metadata, parallel, workers){
  # set multiprocessing for windows
  if(.Platform$OS.type == "windows"){
    bpparam <- BiocParallel::SnowParam(workers = workers)
  }
  # set multiprocessing for linux
  if(.Platform$OS.type == "unix"){
    bpparam <- BiocParallel::MulticoreParam(workers = workers)
  }
  # tidy check count- and metadata
  if(tidy_check(countdata = countdata, metadata = metadata)){
    # create deseqdataset
    out <- DESeqDataSetFromMatrix(countData = data.frame(countdata),
                                  colData = metadata,
                                  design = ~ experimentCondition,
                                  tidy = TRUE)
    # cpm normalization
    sizeFactors(out) <- colSums(column_to_rownames(countdata, var = "gene_symbol"))/1E6
  }
  # start time for DESeq
  start <- Sys.time()
  # run DESeq
  out <- DESeq(object = out,
               parallel = parallel,
               BPPARAM = bpparam)
  # end time for DESeq
  end <- Sys.time()
  # print time difference
  print(difftime(end, start, units="auto"))
  # return out object (dds)
  return(out)
}
```


### runContrast

```{r runContrast}
runContrast <- function(dds, contrast, parallel, workers){
  
  # set multiprocessing for windows
  if(.Platform$OS.type == "windows"){
    bpparam <- BiocParallel::SnowParam(workers = workers)
  }
  # set multiprocessing for linux
  if(.Platform$OS.type == "unix"){
    bpparam <- BiocParallel::MulticoreParam(workers = workers)
  }
  
  res <- data.frame()
  
  start <- Sys.time()
  
  for (i in 1:nrow(contrast)){
    
    experimentCondition <- contrast[i, ]$experimentCondition
    controlCondition <- contrast[i, ]$controlCondition
    
    current_contrast <-  c("experimentCondition", experimentCondition, controlCondition)
    
    cat("run: ", sprintf("%03d", i),"/", nrow(contrast),": ", experimentCondition, "vs", controlCondition, "\n")
    
    res_temp <- DESeq2::results(dds,
                                parallel = parallel,
                                BPPARAM = bpparam,
                                contrast = current_contrast,
                                pAdjustMethod = "BH")
    
    res <- res |> 
      bind_rows(data.frame(res_temp) |> 
                  rownames_to_column(var = "gene_symbol") |> 
                  mutate(experimentCondition = experimentCondition, 
                         controlCondition = controlCondition,
                         shrinkage = "no"))
    
    res_normal <- DESeq2::lfcShrink(dds,
                                    parallel = parallel,
                                    BPPARAM = bpparam,
                                    contrast = current_contrast,
                                    type = "normal")
    res <- res |> 
      bind_rows(data.frame(res_normal) |> 
                  rownames_to_column(var = "gene_symbol") |> 
                  mutate(experimentCondition = experimentCondition, 
                         controlCondition = controlCondition,
                         shrinkage = "normal"))
    
    res_ashr <- DESeq2::lfcShrink(dds,
                                  parallel = parallel,
                                  BPPARAM = bpparam,
                                  contrast = current_contrast,
                                  type = "ashr")
    
    res <- res |> 
      bind_rows(data.frame(res_ashr) |> 
                  rownames_to_column(var = "gene_symbol") |> 
                  mutate(experimentCondition = experimentCondition, 
                         controlCondition = controlCondition,
                         shrinkage = "ashr"))
  } # end contrast loop
  
  # end time for loop contrast
  end <- Sys.time()
  # print time difference
  print(difftime(end, start, units="auto"))
  cat("\n")
  
  return(res)
}
```

### relevanceFilter

```{r relevanceFilter}
relevanceFilter <- function(countdata, metadata, cpm_threshold = 1){
  # countdata is a dataframe with probe_id as rownames and columns as samples
  # metadata is a dataframe containing id and condition columns
  countdata[countdata >= cpm_threshold] <- 1
  countdata[countdata < cpm_threshold] <- 0
  
  countdata <- melt.data.table(data = countdata |> 
                                 rownames_to_column(var = "probe") |> 
                                 as.data.table(),
                               id.vars = "probe",
                               measure.vars = colnames(countdata),
                               variable.name = "id",
                               value.name = "count")
  # bind metadata columns id and experimentCondition
  countdata <- countdata |>
    left_join(y = metadata |> 
                select(id, experimentCondition),
              by = "id")
  
  # number samples per condition group
  output <- countdata |>
    group_by(probe, experimentCondition) |>
    summarise(count_sum = sum(count), .groups = "drop") |>
    left_join(
      y = countdata |>
        distinct(id, experimentCondition) |>
        group_by(experimentCondition) |>
        mutate(n_condition = n()) |>
        distinct(experimentCondition, n_condition), 
      by = "experimentCondition"
    ) |>
    ungroup() |>
    mutate(relevanceFilterPass = if_else(
      condition = count_sum >= n_condition * 0.75, 
      true = TRUE, 
      false = FALSE)) |>
    group_by(probe) |>
    summarise(relevanceFilterPass = sum(relevanceFilterPass), .groups = "drop") |>
    mutate(relevanceFilterPass = if_else(
      relevanceFilterPass >= 1, 
      true = TRUE, 
      false = FALSE)
    )
}
```

### getHumanNCBI

```{r getHumanNCBI}
getHumanNCBI <- function(){
  obj <- data.table::fread(input = "https://ftp.ncbi.nlm.nih.gov/gene/DATA/GENE_INFO/Mammalia/Homo_sapiens.gene_info.gz", 
                           stringsAsFactors = FALSE, 
                           fill = TRUE)
  
  obj <- data.frame(obj)
  obj <- obj[c("Symbol","GeneID","dbXrefs","Modification_date")]
  obj$dbXrefs <- paste0(obj$dbXrefs, "|")
  
  data.frame(
    gene_symbol = paste0("id_", obj$Symbol),
    entrez_id = paste0("id_", obj$GeneID),
    ensembl_id = paste0("id_", stringr::str_extract(str = obj$dbXrefs, "(?<=embl:)(.*?)(?=\\|)")),
    hgnc_id = paste0("id_", stringr::str_extract(string = obj$dbXrefs, "(?<=HGNC:HGNC:)(.*?)(?=\\|)")),
    mim_id = paste0("id_", stringr::str_extract(string = obj$dbXrefs, "(?<=MIM:)(.*?)(?=\\|)")),
    alliancegenome_id = paste0("id_", stringr::str_extract(string = obj$dbXrefs, "(?<=AllianceGenome:HGNC:)(.*?)(?=\\|)")),
    modification_date = obj$Modification_date,
    user_query_date = format(Sys.time(), "%Y%m%d")
  )
}
```

### updateProbe

Get updated gene symbols from NCBI gene database to replace the gene symbol in the probe_name from the manifest

```{r updateProbe}
updateProbe <- function(probe_manifest, countdata, human_ncbi){
  
  # check countdata
  stopifnot(exprs = {has_rownames(countdata)})
  
  # check probe_manifest
  stopifnot(exprs = {"probe_id" %in% colnames(probe_manifest)})
  stopifnot(exprs = {"entrez_id" %in% colnames(probe_manifest)})
  stopifnot(exprs = {"gene_symbol" %in% colnames(probe_manifest)})
  
  # join probe_id from countdata to probe_manifest
  df <- data.frame(probe_name = rownames(countdataRaw)) |> 
    left_join(y = probe_manifest, 
              by = join_by(probe_name))
  
  # match with ncbi through entrez_id
  df$new_gene_symbol <- human_ncbi$gene_symbol[match(df$entrez_id,human_ncbi$entrez_id)]
  df$new_entrez_id <- human_ncbi$entrez_id[match(df$entrez_id,human_ncbi$entrez_id)]
  
  
  # match with ncbi through ensembl_id
  df$new_gene_symbol <- ifelse(is.na(df$new_gene_symbol), 
                               yes = human_ncbi$gene_symbol[match(df$ensembl_id,human_ncbi$ensembl_id)], 
                               no = df$new_gene_symbol)
  
  df$new_entrez_id <- ifelse(is.na(df$new_entrez_id), 
                             yes = human_ncbi$entrez_id[match(df$ensembl_id,human_ncbi$ensembl_id)], 
                             no = df$new_entrez_id)
  
  # match with ncbi through gene_symbol
  df$new_gene_symbol <- ifelse(is.na(df$new_gene_symbol), 
                               yes = human_ncbi$gene_symbol[match(df$gene_symbol,human_ncbi$gene_symbol)], 
                               no = df$new_gene_symbol)
  
  df$new_entrez_id <- ifelse(is.na(df$new_entrez_id), 
                             yes = human_ncbi$entrez_id[match(df$gene_symbol,human_ncbi$gene_symbol)], 
                             no = df$new_entrez_id)
  df
}
```

# Import {.tabset .tabset-fade .tabset-pills}

## probeManifest

```{r Import-probeManifest}
probeManifest <- read.table("DATA/probeManifest_Human_1.2.txt", header = TRUE, sep = "\t"); print(data)

```

## countdataRaw

```{r Import-countdataRaw}
countdataRaw <- read_csv("DATA/countdata")
probe_id_columns <- grep("probe_id", names(countdataRaw))
countdataRaw <- countdataRaw[, -probe_id_columns[-1]]


```

## metadata

```{r Import-metadata}
metadata <- read_csv("DATA/File_S5_httr_mcf7_screen_qc_metrics (1).csv")

```

## humanNCBI

```{r Import-humanNCBI}
humanNCBI <- getHumanNCBI()
```

# Wrangle {.tabset .tabset-fade .tabset-pills}

## probeManifest

```{r Wrangle-probeManifest}
probeManifest <- probeManifest |> 
  clean_names() |> 
  select(probe_id,
         gene_symbol, 
         probe_name,
         entrez_id) |> 
  distinct() |> 
  drop_na()
  
  # ensembl_id = "ensembl_gene_id") |> 
```

## humanNCBI

```{r NCBI-gene}
humanNCBI <- humanNCBI |> 
  mutate(gene_symbol = str_remove(gene_symbol, "id_"),
         entrez_id = str_remove(entrez_id, "id_"),
         ensembl_id  = str_remove(ensembl_id, "id_")
  ) |> 
  distinct(gene_symbol, entrez_id, ensembl_id)
```

## countdataRaw

```{r Wrangle-countdataRaw}
countdataRaw <- countdataRaw |> 
  column_to_rownames(var = "probe_id")

if(file.exists(file.path(outputPath, glue("{params$project}_DAM_updateProbeResult.txt")))){
  updateProbeResult <- fread(file = file.path(outputPath, glue("{params$project}_DAM_updateProbeResult.txt"))) |> tibble()
} else {
  updateProbeResult <- updateProbe(
    probe_manifest = probeManifest, 
    countdata = countdataRaw, 
    human_ncbi = humanNCBI
  )
}

y <- updateProbeResult |> 
  filter(!is.na(new_gene_symbol) & !is.na(new_entrez_id)) |> 
  mutate(new_probe_name = paste0(new_gene_symbol, "_", probe_id)) |> 
  select(probe_name, new_probe_name)

countdataRaw <- countdataRaw[y$probe_name, ] |> 
  rownames_to_column(var = "probe_name") |> 
  left_join(y = y, 
            by = join_by(probe_name)) |> 
  relocate(where(is.character), .before = everything()) |> 
  select(-probe_name) |> 
  column_to_rownames(var = "new_probe_name")
```

## metadata

```{r Wrangle-metadata}
metadata <- metadata |> 
  mutate(chem_name = ifelse(stype == "vehicle control" & is.na(chem_name), "DMSO", chem_name)) |>
  dplyr::rename(group = stype) |>
  dplyr::rename(id = sample_id) |>
  dplyr::rename(time_level = timeh) |>
  dplyr::rename(concentration_level = dose_level) |>
  mutate(experiment = str_replace_all(chem_name, "[^a-zA-Z0-9 ]", "")) |>
  mutate(experiment = case_when(
    trt_name == "UHRR" ~ "UHRR", 
    trt_name == "SIRO" ~ "sirolimus", 
    trt_name == "TSA" ~ "trichostatin A", 
    trt_name == "HBRR" ~ "HBRR",
    trt_name == "GEN" ~ "genistein", 
    trt_name == "BL DMSO" ~ "BL DMSO",
    trt_name == "BL TSA" ~ "BL TSA",
    TRUE ~ experiment  
  )) |> 
 mutate(group = dplyr::recode(group, "test sample" = "experiment"),
        control = "DMSO") |>
  mutate(group = if_else(group == "QC sample", "MAQC", group)) |>
  mutate(isSequencingControl = if_else(condition = group == "MAQC", true = TRUE, false = FALSE)) |>
 mutate(experimental_plate = pg_id, time = time_level, cell = "MCF7")
  
        
  
  
```

## contrastTable

controlCondition and experimentCondition (previously meanID)

```{r ConditionIdentifiers}
# controlCondition
metadata <- metadata |>
  left_join(
    y = metadata |>
      filter(group %in% c("vehicle control")) |>
      select(experiment, time_level, concentration_level, pg_id) |>
      mutate(controlCondition = paste0("pg",pg_id,"_", gsub("[^[:alnum:]]", "", experiment),"_T",time_level,"_C", concentration_level)) |>
      distinct(control = experiment, time_level, controlCondition, pg_id),
    by = join_by(pg_id, control, time_level)
  )

# experimentCondition
metadata <- metadata |>
  mutate(experimentCondition = paste0("pg",pg_id,"_", gsub("[^[:alnum:]]", "", experiment),"_T",time_level,"_C", concentration_level))

contrastTable <- metadata |>
  distinct(experiment, pg_id,group, experimentCondition, controlCondition) |>
  filter(group %in% c("experiment", "vehicle control", "reference chemical"))


contrastTable
```


# Analyse {.tabset .tabset-fade .tabset-pills}

## sampleSize

Sample filter

```{r Analyse-sampleSize}
sampleSize <- data.frame(
  sampleSize = apply(
    X = countdataRaw,
    MARGIN = 2, 
    FUN = sum
  )
) |> 
  rownames_to_column(
    var = "id"
  ) |> 
  mutate(
    sampleSizeFilter = 3E5, 
    sampleSizeFilterPass = if_else(
      condition = sampleSize >= sampleSizeFilter, 
      true = TRUE, 
      false = FALSE
    ),
    sampleSizeBin = cut(sampleSize, breaks = c(0, 3E5, seq(1E6, max(sampleSize)+1E6, 1E6))
    )
  ) |> 
  tibble() |>
  mutate(id = str_replace_all(id, "-", "."))

# bind sampleSize to metadata
metadata <- metadata |> 
  left_join(
    y = sampleSize, 
    by = join_by(id))
```

## countdataCpm

```{r Analyse-countdataCpm}
countdataCpm <- data.frame(
  apply(
    X = countdataRaw,
    MARGIN = 2, 
    FUN = function(x){x/sum(x)*1E6}
  )
)
```

## relevanceFilter

probe filtering

```{r Analyse-relevanceFilter}
includedSamplesRelevanceFilter <- metadata |>
  filter(sampleSizeFilterPass == TRUE) |>
  filter(isSequencingControl == FALSE) |>
  pull(id)

relevanceFilterOutput <- relevanceFilter(
  countdata = 
    countdataCpm |> 
    select(all_of(includedSamplesRelevanceFilter)), 
  metadata = 
    metadata |> 
    filter(id %in% includedSamplesRelevanceFilter) |> 
    select(id, experimentCondition)
)
cat("Relevance filter pass: \n")
table(relevanceFilterOutput$relevanceFilterPass)
```

## sumProbeCount

Aggregate probe count

```{r Analyse-sumProbeCount}
countdataRawSum <- countdataRaw |> 
  rownames_to_column(var = "probe") |> 
  filter(probe %in% 
           (relevanceFilterOutput |> 
              filter(relevanceFilterPass == TRUE) |> 
              pull(probe))
  ) |> 
  separate(col = probe, into = c("gene_symbol", "discard"), sep = "_") |> 
  group_by(gene_symbol) |> 
  summarise(across(where(is.numeric), sum), .groups = "drop") |> 
  column_to_rownames(var = "gene_symbol")
```

## countdataCpmSum

```{r Analyse-countdataCpmSum}
countdataCpmSum <- data.frame(
  apply(
    X = countdataRawSum,
    MARGIN = 2, 
    FUN = function(x){x/sum(x)*1E6}
  )
)
```

## replicateCorrelation

```{r Analyse-replicateCorrelation}
# Select a threshold for the replicate correlation:
threshold_rep_cor <- 0.9

# Select samples to include for the R-ODAF relevance filter
# includedSamplesReplicateCorrelation <- metadata |>
#   # filter(percentageMappedFilterPass == TRUE) |>
#   filter(sampleSizeFilterPass == TRUE) |>
#   filter(isSequencingControl == FALSE) |>
#   pull(id)
# 
# replicateCorrelationOutput <- countdataCpmSum |> 
#   rownames_to_column(var = "gene_symbol") |> 
#   select(gene_symbol, all_of(includedSamplesReplicateCorrelation)) |> 
#   pivot_longer(cols = where(is.numeric), names_to = "id", values_to = "cpm_count") |> 
#   left_join(y = metadata |> 
#               filter(id %in% includedSamplesReplicateCorrelation) |>
#               select(id, experimentCondition) |> 
#               drop_na(), 
#             by = join_by(id))
# 
# replicateCorrelationOutput <- replicateCorrelationOutput |>
#   select(gene_symbol, id, cpm_count, experimentCondition) |> 
#   nest_by(experimentCondition) |> 
#   mutate(cor = list(
#     data |> 
#       pivot_wider(names_from = id, values_from = cpm_count) |>
#       column_to_rownames(var = "gene_symbol") |>
#       mutate(mean = rowMeans(across(where(is.numeric)))) |>
#       cor() |>
#       data.frame() |>
#       rownames_to_column(var = "id") |>
#       pivot_longer(cols = -id) |>
#       filter(name == "mean") |>
#       filter(id != name))) |> 
#   select(-data) |> 
#   unnest(cols = c(cor)) |> 
#   ungroup() |> 
#   select(-experimentCondition)
# 
# replicateCorrelationOutput <- replicateCorrelationOutput |> 
#   select(-name, id, replicateCorrelation = value) |>
#   mutate(replicateCorrelationFilter = threshold_rep_cor,
#          replicateCorrelationFilterPass = if_else(replicateCorrelation >= replicateCorrelationFilter, TRUE, FALSE))
# 
# metadata <- metadata |> 
#   left_join(replicateCorrelationOutput, by = join_by(id))
```

## replicateCorrelation - log2CPM

```{r Analyse-replicateCorrelation-log2CPM}
# Select samples to include for the R-ODAF relevance filter
includedSamplesReplicateCorrelation <- metadata |>
  # filter(percentageMappedFilterPass == TRUE) |>
  filter(sampleSizeFilterPass == TRUE) |>
  filter(isSequencingControl == FALSE) |>
  pull(id)

replicateCorrelationOutput_log2CPM <- log2(countdataCpmSum+1) |> 
  rownames_to_column(var = "gene_symbol") |> 
  select(gene_symbol, all_of(includedSamplesReplicateCorrelation)) |> 
  pivot_longer(cols = where(is.numeric), names_to = "id", values_to = "cpm_count") |> 
  left_join(y = metadata |> 
              filter(id %in% includedSamplesReplicateCorrelation) |>
              select(id, experimentCondition) |> 
              drop_na(), 
            by = join_by(id))

replicateCorrelationOutput_log2CPM <- replicateCorrelationOutput_log2CPM |>
  select(gene_symbol, id, cpm_count, experimentCondition) |> 
  nest_by(experimentCondition) |> 
  mutate(cor = list(
    data |> 
      pivot_wider(names_from = id, values_from = cpm_count) |>
      column_to_rownames(var = "gene_symbol") |>
      mutate(mean = rowMeans(across(where(is.numeric)))) |>
      cor() |>
      data.frame() |>
      rownames_to_column(var = "id") |>
      pivot_longer(cols = -id) |>
      filter(name == "mean") |>
      filter(id != name))) |> 
  select(-data) |> 
  unnest(cols = c(cor)) |> 
  ungroup() |> 
  select(-experimentCondition)

replicateCorrelationOutput_log2CPM <- replicateCorrelationOutput_log2CPM |> 
  select(-name, id, replicateCorrelation_log2CPM = value) |>
  mutate(replicateCorrelationFilter = threshold_rep_cor,
         replicateCorrelationFilterPass_log2CPM = if_else(replicateCorrelation_log2CPM >= replicateCorrelationFilter, TRUE, FALSE))

metadata <- metadata |> 
  left_join(replicateCorrelationOutput_log2CPM, by = join_by(id)) |>
  mutate(replicateCorrelation_bin = cut(replicateCorrelation_log2CPM, breaks = c(0, 0.7, seq(0.75, 1, 0.05))))
```

```{r}
# Identify outlier replicates
metadata |> as.data.frame() |>
  group_by(cell, experiment, compound_abbr, time, concentration, concentration_level, experimentCondition) |>
  identify_outliers("replicateCorrelation_log2CPM")

metadata_outliers <- metadata |> as.data.frame() |>
  group_by(cell, experiment, compound_abbr, time, concentration, concentration_level, experimentCondition) |>
  reframe(outlier = outlier(replicateCorrelation_log2CPM,logical = T)) 
```

## customBasemean

```{r customBasemean}
includedSamplesCustomBasemean <- metadata |> 
  filter(sampleSizeFilterPass == TRUE) |>
  filter(isSequencingControl == FALSE) |>
  filter(replicateCorrelationFilterPass_log2CPM == TRUE) |> 
  pull(id)

customBasemean <- data.frame(basemean = 
                               apply(X = countdataCpmSum |> 
                                       select(all_of(includedSamplesCustomBasemean)),
                                     MARGIN = 1,
                                     FUN = mean
                               )
) |> 
  rownames_to_column(var = "gene_symbol") %>% 
  tibble() %>% 
  mutate(gene_symbol = paste0("id_", gene_symbol))
```

## PCAcontrol

```{r Analyse-PCAcontrol}
PCAcontrolData <- log2(countdataCpmSum+1) |>
  t() |>
  data.frame() |>
  rownames_to_column(var = "id") |>
  left_join(y = metadata,
            by = join_by(id)) |>
  filter(group != "experiment")

PCAcontrolObject <- prcomp(
  x = PCAcontrolData |>
    select(-all_of(colnames(metadata)))
)
```

## PCAcontrolExperiment

```{r Analyse-PCAcontrolExperiment}
PCAcontrolExperimentData <- list()
PCAcontrolExperimentObject <- list()
countdataCpmSum_pg <- list()

for(i in metadata$pg_id) {
  metadata_select <- metadata |>
    filter(pg_id == i)
  
  countdataCpmSum_pg[[i]] <- log2(countdataCpmSum |> select(all_of(metadata_select |> pull(id))) + 1)
  
  PCAcontrolExperimentData[[i]] <- countdataCpmSum_pg[[i]] |>
    t() |>
    data.frame() |>
    rownames_to_column(var = "id") |>
    left_join(
      y = metadata,
      by = join_by(id)
    )
  
  PCAcontrolExperimentObject[[i]] <- prcomp(
    x = PCAcontrolExperimentData[[i]] |>
      select(-all_of(colnames(metadata)))  
  )
}

```

## DGE analysis - experiments 1 tm 6 

```{r DGE-analysis}
contrastTable1 <- contrastTable |> filter(pg_id %in% c(1, 2, 3, 4, 5, 6)) |>
   filter(group == "experiment")

colnames(countdataRawSum) <- gsub("-", ".",colnames(countdataRawSum))

includedSamplesDESeq2 <- metadata |>
  # filter(percentageMappedFilterPass == TRUE) |>
  filter(sampleSizeFilterPass == TRUE) |>
  filter(isSequencingControl == FALSE) |>
  filter(replicateCorrelationFilterPass_log2CPM == TRUE) |>
  filter(qc_flag == "OK") |>
  pull(id)

if(file.exists(file.path(outputPath, glue("{params$project}_DAM_Output.txt")))){
  DESeq2Result <- fread(file = file.path(outputPath, glue("{params$project}_DAM_Output.txt")))
} else {
  
DESeq2Result <- data.frame()
  
  for(i in unique(contrastTable1$experiment)){
    
    temp_contrast <- data.frame(contrastTable1) |>
      filter(experiment == i)
    
    temp_metadata <- metadata |>
      filter(id %in% includedSamplesDESeq2) |>
      filter(experimentCondition %in% c(temp_contrast$experimentCondition , temp_contrast$controlCondition)) |>
      mutate(experimentCondition = as.factor(experimentCondition)) |>
      relocate(id, .before = everything())
    
    temp_contrast <- temp_contrast |>
      filter( experimentCondition %in% temp_metadata$experimentCondition)
    
    temp_countdata <- countdataRawSum |>
      rownames_to_column(var = "gene_symbol") |>
      select(gene_symbol, temp_metadata$id)
    
    cat(i,"\n")
    
    dds <- runDESeq(countdata = temp_countdata,
                    metadata = temp_metadata, 
                    parallel = FALSE, 
                    workers = 0)
    
    
    tempDESeq2Result <- runContrast(dds = dds, 
                                    contrast = temp_contrast, 
                                    parallel = FALSE, 
                                    workers = 0)
    
    tempDESeq2Result <- tempDESeq2Result |> 
      mutate(nSamplesDESeqDataSetFromMatrix = dds@colData@nrows)
    
    DESeq2Result <- DESeq2Result |> 
      bind_rows(tempDESeq2Result)
    
    rm(temp_contrast, temp_metadata, temp_countdata, dds, tempDESeq2Result)
  }
}

```



```{r}
contrastTable1 <- contrastTable |> filter(pg_id %in% c(9, 10)) |>
   filter(group == "experiment")
  
if(file.exists(file.path(outputPath, glue("{params$project}_DAM_Output.txt")))){
  DESeq2Result1 <- fread(file = file.path(outputPath, glue("{params$project}_DAM_Output.txt")))
} else {
  
DESeq2Result1 <- data.frame()
  
  for(i in unique(contrastTable1$experiment)){
    
    temp_contrast <- data.frame(contrastTable1) |>
      filter(experiment == i)
    
    temp_metadata <- metadata |>
      filter(id %in% includedSamplesDESeq2) |>
      filter(experimentCondition %in% c(temp_contrast$experimentCondition , temp_contrast$controlCondition)) |>
      mutate(experimentCondition = as.factor(experimentCondition)) |>
      relocate(id, .before = everything())
    
    temp_contrast <- temp_contrast |>
      filter( experimentCondition %in% temp_metadata$experimentCondition)
    
    temp_countdata <- countdataRawSum |>
      rownames_to_column(var = "gene_symbol") |>
      select(gene_symbol, temp_metadata$id)
    
    cat(i,"\n")
    
    dds <- runDESeq(countdata = temp_countdata,
                    metadata = temp_metadata, 
                    parallel = FALSE, 
                    workers = 0)
    
    
    tempDESeq2Result <- runContrast(dds = dds, 
                                    contrast = temp_contrast, 
                                    parallel = FALSE, 
                                    workers = 0)
    
    tempDESeq2Result <- tempDESeq2Result |> 
      mutate(nSamplesDESeqDataSetFromMatrix = dds@colData@nrows)
    
    DESeq2Result1 <- DESeq2Result1 |> 
      bind_rows(tempDESeq2Result)
    
    rm(temp_contrast, temp_metadata, temp_countdata, dds, tempDESeq2Result)
  }
}
```

##DEG analyse pg 11-13 
```{r}
contrastTable2 <- contrastTable |> filter(pg_id %in% c(11, 12, 13, 14, 15)) |>
   filter(group == "experiment")
  
if(file.exists(file.path(outputPath, glue("{params$project}_DAM_Output.txt")))){
  DESeq2Result1 <- fread(file = file.path(outputPath, glue("{params$project}_DAM_Output.txt")))
} else {
  
DESeq2Result1 <- data.frame()
  
  for(i in unique(contrastTable2$experiment)){
    
    temp_contrast <- data.frame(contrastTable2) |>
      filter(experiment == i)
    
    temp_metadata <- metadata |>
      filter(id %in% includedSamplesDESeq2) |>
      filter(experimentCondition %in% c(temp_contrast$experimentCondition , temp_contrast$controlCondition)) |>
      mutate(experimentCondition = as.factor(experimentCondition)) |>
      relocate(id, .before = everything())
    
    temp_contrast <- temp_contrast |>
      filter( experimentCondition %in% temp_metadata$experimentCondition)
    
    temp_countdata <- countdataRawSum |>
      rownames_to_column(var = "gene_symbol") |>
      select(gene_symbol, temp_metadata$id)
    
    cat(i,"\n")
    
    dds <- runDESeq(countdata = temp_countdata,
                    metadata = temp_metadata, 
                    parallel = FALSE, 
                    workers = 0)
    
    
    tempDESeq2Result <- runContrast(dds = dds, 
                                    contrast = temp_contrast, 
                                    parallel = FALSE, 
                                    workers = 0)
    
    tempDESeq2Result <- tempDESeq2Result |> 
      mutate(nSamplesDESeqDataSetFromMatrix = dds@colData@nrows)
    
    DESeq2Result1 <- DESeq2Result1 |> 
      bind_rows(tempDESeq2Result)
    
    rm(temp_contrast, temp_metadata, temp_countdata, dds, tempDESeq2Result)
  }
}


```



# Visualise {.tabset .tabset-fade .tabset-pills}

## sampleSize

```{r Visualise-sampleSize}
# SampleSize settings
SampleSize_expected <- 300000
SampleSize_threshold_low <- 200000
SampleSize_threshold_high <- 1000000

# Histogram sampleSize color by experimental_plate
(
  sampleSizePlot1 <- ggplot(data = metadata |> 
                              mutate(experimental_plate = factor(experimental_plate, 
                                                                 sort(unique(experimental_plate)))),
                            aes(x = sampleSizeBin, 
                                fill = experimental_plate, 
                                group = experimental_plate)) + 
    geom_histogram(stat = "count") + 
    coord_flip() +
    theme_bw(base_size = 15)
)

# Histogram sampleSize color by time
# (
#   sampleSizePlot2 <- ggplot(data = metadata |> 
#                               mutate(time = factor(time, 
#                                                    sort(unique(time)))),
#                             aes(x = sampleSizeBin, 
#                                 fill = time, 
#                                 group = time)) + 
#     geom_histogram(stat = "count") + 
#     coord_flip() +
#     theme_bw(base_size = 15)
# )


## QC: Sample/library size distribution plot 
QC_column  <- c("experimentCondition", "experiment")

for (plate in unique(metadata$experimental_plate)) {
  plate_data <- metadata %>% filter(experimental_plate == plate)
  
  for (i in QC_column) {
    plate_data <- plate_data |>
      mutate(
        concentration_level = ordered(concentration_level),
        time_level = ordered(time_level),
        VariableQCs = plate_data |> pull(i)
      )
    
    graph_width <- 10 + length(unique(plate_data$VariableQCs)) / 10
    text_x <- round(4 + 300 / length(unique(plate_data$VariableQCs)))
    
    plot <- ggplot(plate_data, aes(x = reorder(VariableQCs, sampleSize, FUN = median), y = sampleSize)) +
      geom_boxplot(size = 0.3, outlier.size = 0.5) +
      geom_point(aes(color = concentration_level), size = 0.5) +
      scale_y_log10(limits = c(1000, max(plate_data$sampleSize, na.rm = TRUE))) +
      geom_hline(yintercept = c(SampleSize_threshold_low, SampleSize_threshold_high, median(plate_data$sampleSize, na.rm = TRUE), SampleSize_expected), 
                 color = c("red", "orange", "green", "darkgreen")) +
      theme_classic() +
      theme(plot.title = element_text(size = 20, face = "bold", vjust = 2, hjust = 0.5), 
            axis.title.x = element_text(size = 16, vjust = 0.25),
            axis.title.y = element_text(size = 16, vjust = 1),
            axis.text.x = element_text(size = text_x, angle = 90, vjust = 0.5, hjust = 1),
            axis.text.y = element_text(size = 16)) +
      ggtitle(paste("Library size distribution - Plate:", plate)) +
      ylab('Library size') +
      xlab(i)
    
    print(plot)
    
    ggsave(paste0(outputFigures, "/SampleSize_", plate, "_", i, ".pdf"), plot = plot, 
           width = graph_width, height = 10, device = "pdf", limitsize = FALSE)
  }
}


# metadata |> 
#   select(experiment, concentration_level, biological_replicate, sampleSize, sampleSizeFilterPass) |> 
#   filter(sampleSizeFilterPass == FALSE) |> 
#   select(-sampleSizeFilterPass) |> 
#   arrange(experiment)
```


## updateProbeResult

```{r Visualise-updateProbeResult}

# total number of probes in manifest
updateProbeResult |> 
  nrow()

# table of probes not included
updateProbeResult |> 
  filter(is.na(new_entrez_id))

# number of probes not included
updateProbeResult |> 
  filter(is.na(new_entrez_id)) |> 
  nrow()

# table of probes included
# table is too large for rmd document
# updateProbeResult |> 
#   filter(!is.na(new_entrez_id))

# number of probes included
updateProbeResult |> 
  filter(!is.na(new_entrez_id)) |> 
  nrow()
```

## relevanceFilter 

```{r Visualise-relevanceFilter}
# n probes that passed the relevance filter
relevanceFilterOutput |> 
  filter(relevanceFilterPass == TRUE) |> 
  nrow()

# n probes that did not pass the relevance filter
relevanceFilterOutput |> 
  filter(relevanceFilterPass == FALSE) |> 
  nrow()

# n genes after probe count sum per gene
length(rownames(countdataRawSum))
```


##replicate correlation 
```{r}
QC_column  <- c("experimentCondition", "experiment")

for (plate in unique(metadata$experimental_plate)) {
  
  # Subset the data for the current experimental_plate
  plate_data <- metadata %>% filter(experimental_plate == plate)
  
  for(i in QC_column) {
    plate_data <- plate_data |>
      mutate(
        concentration_level = ordered(concentration_level),
        time_level = ordered(time_level),  
        VariableQCs = plate_data |> pull(i)
      )
    
    # Bereken de afmetingen en tekstgrootte
    graph_width <- min(10 + length(unique(plate_data$VariableQCs)) / 10, 20)  # Max breedte is 20
    text_x <- round(4 + 300 / length(unique(plate_data$VariableQCs)))
    
    # Maak en print de plot voor replicateCorrelation_log2CPM
    plot <- ggplot(plate_data, aes(x = reorder(VariableQCs, replicateCorrelation_log2CPM, FUN = median), y = replicateCorrelation_log2CPM)) +
      geom_boxplot(size = 0.3, outlier.size = 0.5, outlier.colour = "white", na.rm = TRUE) +  # Exclude NA values
      geom_point(aes(color = concentration_level), size = 1, na.rm = TRUE) +  # Exclude NA values
      geom_hline(yintercept = c(0.8, 0.9, 0.95), color = c("red", "orange", "green")) +
      theme_bw() +
      theme(plot.title = element_text(size = 20, face = "bold", vjust = 2, hjust = 0.5), 
            axis.title.x = element_text(size = 16, vjust = 0.25),
            axis.title.y = element_text(size = 16, vjust = 1),
            axis.text.x = element_text(size = text_x, angle = 90, vjust = 0.5, hjust = 1),
            axis.text.y = element_text(size = 16)) +
      ggtitle("Pearson R replicate correlation") + 
      ylab('Replicate correlation') + 
      xlab(i)
    
    print(plot)
    
    # Bewaar de plot
    ggsave(paste0(outputFigures, "/Replicate_correlation_", plate, "_", i, ".pdf"), plot = plot, 
           width = graph_width, height = 10, device = "pdf", limitsize = FALSE)
  }
  
  for(i in QC_column) {
    # Mutate voor replicateCorrelation_log2CPM (inclusief time_level)
    plate_data <- plate_data |>
      mutate(
        concentration_level = ordered(concentration_level),
        time_level = ordered(time_level),  # Belangrijk: time_level blijft hier
        VariableQCs = plate_data |> pull(i)
      )
    
    # Bereken de afmetingen en tekstgrootte
    graph_width <- min(10 + length(unique(plate_data$VariableQCs)) / 10, 20)  # Max breedte is 20
    text_x <- round(4 + 300 / length(unique(plate_data$VariableQCs)))
    
    # Maak en print de plot voor replicateCorrelation_log2CPM
    plot_log2CPM <- ggplot(plate_data, aes(x = reorder(VariableQCs, replicateCorrelation_log2CPM, FUN = median), y = replicateCorrelation_log2CPM)) +
      geom_boxplot(size = 0.3, outlier.size = 0.5, outlier.colour = "white", na.rm = TRUE) +  # Exclude NA values
      geom_point(aes(color = concentration_level), size = 1, na.rm = TRUE) +  # Exclude NA values
      geom_hline(yintercept = c(0.8, 0.9, 0.95), color = c("red", "orange", "green")) +
      theme_bw() +
      theme(plot.title = element_text(size = 20, face = "bold", vjust = 2, hjust = 0.5), 
            axis.title.x = element_text(size = 16, vjust = 0.25),
            axis.title.y = element_text(size = 16, vjust = 1),
            axis.text.x = element_text(size = text_x, angle = 90, vjust = 0.5, hjust = 1),
            axis.text.y = element_text(size = 16)) +
      ggtitle("Pearson R replicate correlation log2CPM") + 
      ylab('Replicate correlation log2CPM') + 
      xlab(i)
    
    print(plot_log2CPM)
    
    # Bewaar de plot voor replicateCorrelation_log2CPM
    ggsave(paste0(outputFigures, "/Replicate_correlation_log2CPM_", plate, "_", i, ".pdf"), plot = plot_log2CPM, 
           width = graph_width, height = 10, device = "pdf", limitsize = FALSE)
  }
}


```


## PCAcontrol

```{r Visualise-PCAcontrol, eval=FALSE, warning=FALSE, include=FALSE}
# PCAcontrol color by group
(
  PCAcontrolPlot1 <- autoplot(
    object = PCAcontrolObject,
    data = PCAcontrolData,
    color = "group",
    size = 3) +
    theme_bw(base_size = 15)
  
)

# PCAcontrol color by experimental_plate
(
  PCAcontrolPlot2 <- autoplot(
    object = PCAcontrolObject,
    data = PCAcontrolData |>
      mutate(experimental_plate = factor(experimental_plate,
                                         sort(unique(experimental_plate)))),
    color = "experimental_plate",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrol color by time
# (  
#   PCAcontrolPlot3 <- autoplot(
#     object = PCAcontrolObject,
#     data = PCAcontrolData |> 
#       mutate(time = factor(time,
#                            sort(unique(time)))),
#     color = "time",
#     size = 3) +
#     theme_bw(base_size = 15)
)

# PCAcontrol color by sampleSizeFilterPass
(
  PCAcontrolPlot4 <- autoplot(
    object = PCAcontrolObject,
    data = PCAcontrolData,
    color = "sampleSizeFilterPass",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrol color by sampleSizeBin
(
  PCAcontrolPlot5 <- autoplot(
    object = PCAcontrolObject,
    data = PCAcontrolData,
    color = "sampleSizeBin",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrolExperiment color by plate_id from sequencing
(
  PCAcontrolPlot6 <- autoplot(
    object = PCAcontrolObject,
    data = PCAcontrolData,
    color = "plate_id",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrolExperiment color by well_id from sequencing
# (
#   PCAcontrolPlot7 <- autoplot(
#     object = PCAcontrolObject,
#     data = PCAcontrolData,
#     color = "well_id",
#     size = 3) +
#     theme_bw(base_size = 15)
# )

# PCAcontrol color by concentration_level
(  
  PCAcontrolPlot8 <- autoplot(
    object = PCAcontrolObject,
    data = PCAcontrolData |> 
      mutate(time = factor(time,
                           sort(unique(time)))),
    color = "concentration_level",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrol color by biological_replicate
(  
  PCAcontrolPlot9 <- autoplot(
    object = PCAcontrolObject,
    data = PCAcontrolData |> 
      mutate(time = factor(time,
                           sort(unique(time)))),
    color = "biological_replicate",
    size = 3) +
    theme_bw(base_size = 15)
)


# PCAcontrol color by replicateCorrelation_bin
(  
  PCAcontrolPlot10 <- autoplot(
    object = PCAcontrolObject,
    data = PCAcontrolData |> 
      mutate(time = factor(time,
                           sort(unique(time)))),
    color = "replicateCorrelation_bin",
    size = 3) +
    theme_bw(base_size = 15)
)
```

## PCAcontrolExperiment

```{r Visualise-PCAcontrolExperiment, eval=FALSE, include=FALSE}
# PCAcontrolExperiment color by group
(
  PCAcontrolExperimentPlot1 <- autoplot(
    object = PCAcontrolExperimentObject,
    data = PCAcontrolExperimentData,
    color = "group",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrolExperiment color by experimental_plate
(
  PCAcontrolExperimentPlot2 <- autoplot(
    object = PCAcontrolExperimentObject,
    data = PCAcontrolExperimentData |>
      mutate(experimental_plate = factor(experimental_plate,
                                         sort(unique(experimental_plate)))),
    color = "experimental_plate",
    size = 3) +
    theme_bw(base_size = 15)
)

# # PCAcontrolExperiment color by time
# (
#   PCAcontrolExperimentPlot3 <- autoplot(
#     object = PCAcontrolExperimentObject,
#     data = PCAcontrolExperimentData |>
#       mutate(time = factor(time,
#                            sort(unique(time)))),
#     color = "time",
#     size = 3) +
#     theme_bw(base_size = 15)
#   
# )

# PCAcontrolExperiment color by experiment with no legend
(
  PCAcontrolExperimentPlot4 <- autoplot(
    object = PCAcontrolExperimentObject,
    data = PCAcontrolExperimentData |>
      mutate(time = factor(time,
                           sort(unique(time)))),
    color = "experiment",
    size = 3) +
    theme_bw(base_size = 15) +
    theme(legend.position = "none")
)

# PCAcontrolExperiment color by sampleSizeFilterPass
(
  PCAcontrolExperimentPlot5 <- autoplot(
    object = PCAcontrolExperimentObject,
    data = PCAcontrolExperimentData,
    color = "sampleSizeFilterPass",
    size = 3) +
    theme_bw(base_size = 15)
  
)

# PCAcontrolExperiment color by sampleSizeBin
(
  PCAcontrolExperimentPlot6 <- autoplot(
    object = PCAcontrolExperimentObject,
    data = PCAcontrolExperimentData,
    color = "sampleSizeBin",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrolExperiment color by plate_id from sequencing
(
  PCAcontrolExperimentPlot7 <- autoplot(
    object = PCAcontrolExperimentObject,
    data = PCAcontrolExperimentData,
    color = "plate_id",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrolExperiment color by well_id from sequencing
(
  PCAcontrolExperimentPlot8 <- autoplot(
    object = PCAcontrolExperimentObject,
    data = PCAcontrolExperimentData,
    color = "well_id",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrolExperiment color by biological_replicate
(
  PCAcontrolExperimentPlot9 <- autoplot(
    object = PCAcontrolExperimentObject,
    data = PCAcontrolExperimentData,
    color = "biological_replicate",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrolExperiment color by replicateCorrelation_log2CPM
(
  PCAcontrolExperimentPlot10 <- autoplot(
    object = PCAcontrolExperimentObject,
    data = PCAcontrolExperimentData,
    color = "replicateCorrelation_log2CPM",
    size = 3) +
    theme_bw(base_size = 15)
)

# PCAcontrolExperiment color by replicateCorrelation_bin
(
  PCAcontrolExperimentPlot11 <- autoplot(   
    object = PCAcontrolExperimentObject,
    data = PCAcontrolExperimentData,
    color = "replicateCorrelation_bin",
    size = 3) +
    theme_bw(base_size = 15)
)

```


##PCA loop plots 
```{r PCA-loop, eval=FALSE, include=FALSE}

for (i in seq_along(PCAcontrolExperimentObject)) {
  current_data <- PCAcontrolExperimentData[[i]]
  current_object <- PCAcontrolExperimentObject[[i]]
  
  # PCAcontrolExperiment color by group
  PCAcontrolExperimentPlot1 <- autoplot(
    object = current_object,
    data = current_data,
    color = "group", shape = "qc_flag", 
    size = 3
  ) +
  ggtitle(paste("PCA Plot - Color by group for pg [", i, "]")) +  
  theme_bw(base_size = 15) +
  theme(plot.title = element_text(hjust = 0.5))
  
  print(PCAcontrolExperimentPlot1)
  
  ggsave( paste0(outputFigures, "/PCA/PCA_Plot_group_pg_", i, ".pdf"),  plot = PCAcontrolExperimentPlot1, width = 10, height = 8,device = "pdf", limitsize = FALSE)

  
  # PCAcontrolExperiment color by experiment with no legend
  PCAcontrolExperimentPlot4 <- autoplot(
    object = current_object,
    data = current_data |>
      mutate(experiment = factor(experiment,
                                 sort(unique(experiment)))),
    color = "experiment", shape = "qc_flag", 
    size = 3
  ) +
  ggtitle(paste("PCA Plot - Color by experiment for pg [", i, "]")) + 
  theme_bw(base_size = 15) +
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none")
  
  print(PCAcontrolExperimentPlot4)
  ggsave( paste0(outputFigures, "/PCA/PCA_Plot_experiment_pg_", i, ".pdf"),  plot = PCAcontrolExperimentPlot4,  width = 10, height = 8, device = "pdf", limitsize = FALSE)
  
  # PCAcontrolExperiment color by sampleSizeFilterPass
  PCAcontrolExperimentPlot5 <- autoplot(
    object = current_object,
    data = current_data,
    color = "sampleSizeFilterPass", shape = "qc_flag", size = 3
  ) +
  ggtitle(paste("PCA Plot - Color by sampleSizeFilterPass for pg [", i, "]")) +  
  theme_bw(base_size = 15) +
  theme(plot.title = element_text(hjust = 0.5))
  
  print(PCAcontrolExperimentPlot5)
  ggsave(paste0(outputFigures, "/PCA/PCA_Plot_sampleSizeFilterPass_pg_", i, ".pdf"), plot = PCAcontrolExperimentPlot5,width = 10,height = 8,device = "pdf",limitsize = FALSE)

  # PCAcontrolExperiment color by replicateCorrelation_log2CPM
  PCAcontrolExperimentPlot10 <- autoplot(
    object = current_object,
    data = current_data,
    color = "replicateCorrelation_log2CPM",shape = "qc_flag", size = 3
  ) +
  ggtitle(paste("PCA Plot - Color by replicateCorrelation_log2CPM for pg [", i, "]")) + 
  theme_bw(base_size = 15)
  ggsave(paste0(outputFigures, "/PCA/PCA_Plot_replicateCorrelation_log2CPM_pg_", i, ".pdf"), plot = PCAcontrolExperimentPlot10, width = 10, height = 8,device = "pdf", limitsize = FALSE)
  
  # PCAcontrolExperiment color by replicateCorrelation_bin
  PCAcontrolExperimentPlot11 <- autoplot(
    object = current_object,
    data = current_data,
    color = "replicateCorrelation_bin", shape = "qc_flag",
    size = 3
  ) +
  ggtitle(paste("PCA Plot - Color by replicateCorrelation_bin for pg [", i, "]")) + 
  theme_bw(base_size = 15)

  print(PCAcontrolExperimentPlot11)

  ggsave( paste0(outputFigures, "/PCA/PCA_Plot_replicateCorrelation_bin_pg_", i, ".pdf"),  plot = PCAcontrolExperimentPlot11, width = 10,height = 8, device = "pdf", limitsize = FALSE)

  # PCAcontrolExperiment color by sampleSizeBin
  PCAcontrolExperimentPlot6 <- autoplot(
    object = current_object,
    data = current_data,
    color = "sampleSizeBin", shape = "qc_flag",
    size = 3
  ) +
  ggtitle(paste("PCA Plot - Color by sampleSizeBin for pg [", i, "]")) +
  theme_bw(base_size = 15)

  print(PCAcontrolExperimentPlot6)
  ggsave( paste0(outputFigures, "/PCA/PCA_Plot_sampleSizeBin_pg_", i, ".pdf"), plot = PCAcontrolExperimentPlot6, width = 10, height = 8,device = "pdf",limitsize = FALSE )
  
  # PCAcontrolExperiment color by plate_id from sequencing
  PCAcontrolExperimentPlot7 <- autoplot(
    object = current_object,
    data = current_data,
    color = "plate_id", shape = "qc_flag",
    size = 3
  ) +
  ggtitle(paste("PCA Plot - Color by plate_id from sequencing for pg [", i, "]")) +  
  theme_bw(base_size = 15)

  print(PCAcontrolExperimentPlot7)

  ggsave( paste0(outputFigures, "/PCA/PCA_Plot_plate_id_pg_", i, ".pdf"), plot = PCAcontrolExperimentPlot7, width = 10, height = 8, device = "pdf", limitsize = FALSE )
} 

```

## MA plots

```{r Visualise-MAplotShrinkage, eval=FALSE, include=FALSE}
DESeq2Result1 <- DESeq2Result1 |>
  mutate(baseMean_log2 = log2(baseMean + 0.01),
         shrinkage = ordered(shrinkage, levels = c("no", "normal", "ashr")))

baseMean_limits = c(min(DESeq2Result1$baseMean_log2), max(DESeq2Result$baseMean_log2))
padj_limits = c(min(-log10(na.omit(DESeq2Result$padj))), max(-log10(na.omit(DESeq2Result$padj))))
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
color_on = "baseMean"
color_on = "p_adj"

# DESeq2Result_filter <- DESeq2Result |> filter(baseMean_log2 > 0)

sc <- scale_color_gradientn(colours = myPalette(201), limits = c(0,10), breaks = c(0:10), oob = scales::squish)
sf <- scale_fill_gradientn(colours = myPalette(201), limits = c(0,10), breaks = c(0:10), oob = scales::squish)

experiments <- metadata |> 
  filter(group %in% c("experiment", "experiment_control")) |>
  pull(experiment) |> unique()

for(i in experiments) {
  print(i)
  DESeq2Result_subset <- DESeq2Result |>
    left_join(metadata |>
                distinct(experimentCondition, cell, experiment, time_level, concentration_level), by = "experimentCondition") |>
    filter(experiment == i)

  # axis_limits = c(min(DESeq2Result_subset$log2FoldChange), max(DESeq2Result_subset$log2FoldChange))

  p = ggplot(DESeq2Result_subset, aes(x = baseMean_log2,  y = log2FoldChange)) +
        geom_point(aes(color = -log10(padj))) +
        ggtitle(paste0("MA plot of " , i)) +
        xlab("log2 baseMean expression") +
        ylab("log2 fold change") +
        # ylim(axis_limits) +
        xlim(baseMean_limits) +
        facet_grid(shrinkage ~ concentration_level, scales = "fixed") +
        sc + sf


  ggsave(paste0(outputFigures,"/MA_plot/MA_plot_Gene_log2FC_shrinkage_",i,".pdf"), plot = p, width = 16, height = 8, device = "pdf")

  # ggsave(paste0(outputFigures,"/MA_plot/MA_plot_Gene_log2FC_shrinkage_",i,".png"), plot = p, width = 16, height = 8, device = "png")

}



```

##MA_plots_9 & 10 

```{r}
DESeq2Result1 <- DESeq2Result1 |>
  mutate(baseMean_log2 = log2(baseMean + 0.01),
         shrinkage = ordered(shrinkage, levels = c("no", "normal", "ashr")))

baseMean_limits = c(min(DESeq2Result1$baseMean_log2), max(DESeq2Result1$baseMean_log2))
padj_limits = c(min(-log10(na.omit(DESeq2Result1$padj))), max(-log10(na.omit(DESeq2Result1$padj))))
myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
color_on = "baseMean"
color_on = "p_adj"

# DESeq2Result_filter <- DESeq2Result |> filter(baseMean_log2 > 0)

sc <- scale_color_gradientn(colours = myPalette(201), limits = c(0,10), breaks = c(0:10), oob = scales::squish)
sf <- scale_fill_gradientn(colours = myPalette(201), limits = c(0,10), breaks = c(0:10), oob = scales::squish)

experiments <- metadata |> 
  filter(group %in% c("experiment", "experiment_control")) |>
  pull(experiment) |> unique()

for(i in experiments) {
  print(i)
  DESeq2Result_subset <- DESeq2Result1 |>
    left_join(metadata |>
                distinct(experimentCondition, cell, experiment, time_level, concentration_level), by = "experimentCondition") |>
    filter(experiment == i)

  # axis_limits = c(min(DESeq2Result_subset$log2FoldChange), max(DESeq2Result_subset$log2FoldChange))

  p = ggplot(DESeq2Result_subset, aes(x = baseMean_log2,  y = log2FoldChange)) +
        geom_point(aes(color = -log10(padj))) +
        ggtitle(paste0("MA plot of " , i)) +
        xlab("log2 baseMean expression") +
        ylab("log2 fold change") +
        # ylim(axis_limits) +
        xlim(baseMean_limits) +
        facet_grid(shrinkage ~ concentration_level, scales = "fixed") +
        sc + sf


  ggsave(paste0(outputFigures,"/MA_plot/MA_plot_Gene_log2FC_shrinkage_",i,".pdf"), plot = p, width = 16, height = 8, device = "pdf")

  # ggsave(paste0(outputFigures,"/MA_plot/MA_plot_Gene_log2FC_shrinkage_",i,".png"), plot = p, width = 16, height = 8, device = "png")

}

```


## DGE

```{r Visualise-DGE}
threshold_fc <- 1
threshold_padj <- 0.05
shrinkage_method <- "no"

res <- DESeq2Result %>% 
  left_join(y = metadata %>% distinct(experimentCondition, experiment, time, concentration_level), join_by(experimentCondition)) %>% 
  filter(shrinkage == shrinkage_method) %>%
  nest_by(experimentCondition, cell, experiment, time, concentration_level, shrinkage, controlCondition) %>% 
  mutate(data_up = list(data %>% filter(log2FoldChange >= log2(threshold_fc) & padj < threshold_padj)),
         data_down = list(data %>% filter(log2FoldChange <= log2(threshold_fc) & padj < threshold_padj))) %>% 
  mutate(resdeg_up_n = nrow(data_up),
         resdeg_down_n = 1 * nrow(data_down)) %>% 
  select(experimentCondition, cell, experiment, concentration_level, time, shrinkage, resdeg_up_n, resdeg_down_n) %>% 
  pivot_longer(cols = c(resdeg_up_n, resdeg_down_n), values_to = "resdeg") %>% 
  mutate(resdeg_log10_n = log10(resdeg + 1)) %>% 
  mutate(concentration_level = ordered(concentration_level)) %>% 
  group_by(experiment) %>% 
  mutate(sumall = sum(abs(resdeg_log10_n))) %>% 
  mutate(regulation = if_else(name == "resdeg_down_n", "Down", "Up"),
         regulation = ordered(regulation, levels = c("Up", "Down")))

ggplot(data = res, 
       aes(x = reorder(experiment, sumall),
           y = resdeg_log10_n,
           fill = concentration_level)) + 
  geom_bar(stat = "identity", position = position_stack(reverse = TRUE)) +
  coord_flip(ylim = c(0,15)) +
  labs(x = NULL, y = paste0("log10(n+1 gene with |log2fc| > ",log2(threshold_fc)," and padj < ",threshold_padj,")")) +
  facet_grid(cell ~ time + regulation, scales = "free_y", space = "free_y", switch = "y") +
  scale_y_continuous(limits = c(0,15),breaks = c(0,3,6,9,12,15)) +
  # themeFigure(base_size = 20) +
  scale_fill_brewer(palette = "Dark2") +
  theme_bw()

print(
  ggplot(data = res %>%
           mutate(resdeg = if_else(regulation=="Down", resdeg*-1, resdeg)),
         mapping = aes(x = concentration_level,
                       y = resdeg,
                       fill = regulation)) +
    geom_bar(stat = "identity") +
    facet_wrap(experiment ~ time, ncol = 14, scales = "fixed") +
    labs(title = paste0("Number of DEGS with |log2fc| > ",log2(threshold_fc)," and padj < ",threshold_padj,""), x = "Concentration level", y = "n deg") +
    theme_bw() +
  scale_fill_discrete())

ggsave(paste0(outputFigures,"/DEG_plot_no_shrinkage_padj_",threshold_padj,".pdf"), width = 18, height = 12, device = "pdf")

print(
  ggplot(data = res %>%
                mutate(resdeg_log10_n = if_else(regulation=="Down", resdeg_log10_n*-1, resdeg_log10_n)),
         mapping = aes(x = concentration_level,
                       y = resdeg_log10_n,
                       fill = regulation)) +
    geom_bar(stat = "identity") +
    facet_wrap(experiment ~ time, ncol = 14, scales = "fixed") +
    labs(title = paste0("Number of DEGS with |log2fc| > ",log2(threshold_fc)," and padj < ",threshold_padj,""), x = "Concentration level", y = "log10(n deg)") +
    theme_bw() +
  scale_fill_discrete())

ggsave(paste0(outputFigures,"/DEG_plot_log10_no_shrinkage_padj_",threshold_padj,".pdf"), width = 18, height = 12, device = "pdf")


for (i in unique(res$experiment)){

  print(
    ggplot(data = res %>%
             filter(experiment == i)
             # mutate(resdeg = if_else(regulation=="Down", resdeg*-1, resdeg))
           ,
           mapping = aes(x = concentration_level,
                         y = resdeg,
                         fill = regulation)) +
      geom_bar(stat = "identity") +
      facet_wrap(cell ~ time) +
      labs(title = i, x = "Concentration level", y = "n deg") +
      theme_bw() +
  scale_fill_discrete()
  )
  ggsave(paste0(outputFigures,"/DEG_plot/DEG_plot_",i,"_stack_no_shrinkage_padj_",threshold_padj,"_log2fc_",log2(threshold_fc),".pdf"), width = 18, height = 12, device = "pdf")
}
```


## DGE abs log2FC of normal shrinkage

```{r Visualize-DGE-abs-log2fc}
threshold_fc <- 1
threshold_padj <- 0.05

DESeq2Result_log2fc <- DESeq2Result |>
  filter(shrinkage == "normal") |>
  mutate(log2fc_abs = abs(log2FoldChange)) |>
  select(experimentCondition, log2fc_abs, log2FoldChange) |>
  group_by(experimentCondition) |>
  summarise(n_genes = n(),
            log2fc_mean = mean(log2fc_abs, na.rm = T),
            log2fc_median = median(log2fc_abs, na.rm = T),
            log2fc_sd = sd(log2fc_abs, na.rm = T), .groups = "drop") |>
  left_join(metadata |> distinct(experimentCondition, cell, experiment, time, concentration_level), by = "experimentCondition")
  
ggplot(data = DESeq2Result_log2fc,
       mapping = aes(x = concentration_level,
                     y = log2fc_mean)) +
  geom_bar(stat = "identity") +
  facet_wrap(experiment ~ time, ncol = 14, scales = "fixed") +
  labs(title = paste0("Number of DEGS with |log2fc| > ",log2(threshold_fc)," and padj < ",threshold_padj,""), x = "Concentration level", y = "n deg") +
  theme_bw() +
  scale_fill_discrete()

ggsave(paste0(outputFigures,"/DEG_abs_log2fc_normal_shrinkage.pdf"), width = 18, height = 12, device = "pdf")

```


##DGE pg 1 tm 5 
```{r}
threshold_fc <- 1
threshold_padj <- 0.05
shrinkage_method <- "no"

res1 <- DESeq2Result1 %>% 
  left_join(y = metadata %>% distinct(experimentCondition, experiment, time, concentration_level, cell), join_by(experimentCondition)) %>% 
  filter(shrinkage == shrinkage_method) %>%
  nest_by(experimentCondition, cell, experiment, time, concentration_level, shrinkage, controlCondition) %>% 
  mutate(data_up = list(data %>% filter(log2FoldChange >= log2(threshold_fc) & padj < threshold_padj)),
         data_down = list(data %>% filter(log2FoldChange <= log2(threshold_fc) & padj < threshold_padj))) %>% 
  mutate(resdeg_up_n = nrow(data_up),
         resdeg_down_n = 1 * nrow(data_down)) %>% 
  select(experimentCondition, cell, experiment, concentration_level, time, shrinkage, resdeg_up_n, resdeg_down_n) %>% 
  pivot_longer(cols = c(resdeg_up_n, resdeg_down_n), values_to = "resdeg") %>% 
  mutate(resdeg_log10_n = log10(resdeg + 1)) %>% 
  mutate(concentration_level = ordered(concentration_level)) %>% 
  group_by(experiment) %>% 
  mutate(sumall = sum(abs(resdeg_log10_n))) %>% 
  mutate(regulation = if_else(name == "resdeg_down_n", "Down", "Up"),
         regulation = ordered(regulation, levels = c("Up", "Down")))

ggplot(data = res1, 
       aes(x = reorder(experiment, sumall),
           y = resdeg_log10_n,
           fill = concentration_level)) + 
  geom_bar(stat = "identity", position = position_stack(reverse = TRUE)) +
  coord_flip(ylim = c(0,15)) +
  labs(x = NULL, y = paste0("log10(n+1 gene with |log2fc| > ",log2(threshold_fc)," and padj < ",threshold_padj,")")) +
  facet_grid(cell ~ time + regulation, scales = "free_y", space = "free_y", switch = "y") +
  scale_y_continuous(limits = c(0,15),breaks = c(0,3,6,9,12,15)) +
  # themeFigure(base_size = 20) +
  scale_fill_brewer(palette = "Dark2") +
  theme_bw()

print(
  ggplot(data = res1 %>%
           mutate(resdeg = if_else(regulation=="Down", resdeg*-1, resdeg)),
         mapping = aes(x = concentration_level,
                       y = resdeg,
                       fill = regulation)) +
    geom_bar(stat = "identity") +
    facet_wrap(experiment ~ time, ncol = 14, scales = "fixed") +
    labs(title = paste0("Number of DEGS with |log2fc| > ",log2(threshold_fc)," and padj < ",threshold_padj,""), x = "Concentration level", y = "n deg") +
    theme_bw() +
  scale_fill_discrete())

ggsave(paste0(outputFigures,"/DEG_plot_no_shrinkage_padj_11_15",threshold_padj,".pdf"), width = 18, height = 12, device = "pdf")

print(
  ggplot(data = res1 %>%
                mutate(resdeg_log10_n = if_else(regulation=="Down", resdeg_log10_n*-1, resdeg_log10_n)),
         mapping = aes(x = concentration_level,
                       y = resdeg_log10_n,
                       fill = regulation)) +
    geom_bar(stat = "identity") +
    facet_wrap(experiment ~ time, ncol = 14, scales = "fixed") +
    labs(title = paste0("Number of DEGS with |log2fc| > ",log2(threshold_fc)," and padj < ",threshold_padj,""), x = "Concentration level", y = "log10(n deg)") +
    theme_bw() +
  scale_fill_discrete())

ggsave(paste0(outputFigures,"/DEG_plot_log10_no_shrinkage_padj_11_15",threshold_padj,".pdf"), width = 18, height = 12, device = "pdf")


for (i in unique(res1$experiment)){

  print(
    ggplot(data = res1 %>%
             filter(experiment == i)
             # mutate(resdeg = if_else(regulation=="Down", resdeg*-1, resdeg))
           ,
           mapping = aes(x = concentration_level,
                         y = resdeg,
                         fill = regulation)) +
      geom_bar(stat = "identity") +
      facet_wrap(cell ~ time) +
      labs(title = i, x = "Concentration level", y = "n deg") +
      theme_bw() +
  scale_fill_discrete()
  )
  ggsave(paste0(outputFigures,"/DEG_plot/DEG_plot_",i,"_stack_no_shrinkage_padj_",threshold_padj,"_log2fc_",log2(threshold_fc),".pdf"), width = 18, height = 12, device = "pdf")
}
```




# Save {.tabset .tabset-fade .tabset-pills}

## table

```{r table}
DESeq2Result_upload <- DESeq2Result |>
  left_join(y = metadata |> distinct(experimentCondition, cell, experiment, time, concentration, concentration_level), join_by(experimentCondition)) |> 
  left_join(y = updateProbeResult |> distinct(new_gene_symbol, new_entrez_id) |> select(gene_symbol = new_gene_symbol, entrez_id = new_entrez_id), join_by(gene_symbol)) |> 
  mutate(experiment_shrinkage = paste0(experiment, "_", toupper(shrinkage))) |>
  dplyr::select(experiment = experiment_shrinkage,
         time,
         conc = concentration,
         gene_id = entrez_id,
         # gene_symbol,
         log2fc = log2FoldChange,
         pvalue,
         padj) 

metadata <- metadata |>
  select(-plate_sample, -compound_cas_lukas, -check_cas, -conc_ic_check, -VariableQCs)

# relevanceFilterOutput
data.table::fwrite(x = relevanceFilterOutput,
                   file = file.path(outputPath, glue("{params$project}_DAM_relevanceFilterOutput.txt")),
                   sep = "\t",
                   row.names = F)

# updateProbeResult
data.table::fwrite(x = updateProbeResult,
                   file = file.path(outputPath, glue("{params$project}_DAM_updateProbeResult.txt")),
                   sep = "\t",
                   row.names = F)

# DAM DESeq2Result
data.table::fwrite(x = DESeq2Result,
                   file = file.path(outputPath, glue("{params$project}_DAM_output.txt")),
                   sep = "\t",
                   row.names = F)

# DAM metadata
data.table::fwrite(x = metadata,
                   file = file.path(outputPath, glue("{params$project}_DAM_metadata.txt")),
                   sep = "\t",
                   row.names = F)

# customBasemean
data.table::fwrite(x = customBasemean,
                   file = file.path(outputPath, glue("{params$project}_DAM_customBasemean.txt")),
                   sep = "\t",
                   row.names = F)

# TXG-MAPr upload file
data.table::fwrite(x = DESeq2Result_upload,
                   file = file.path(outputPath, glue("{params$project}_DAM_TXG_MAPr_upload.txt")),
                   sep = "\t",
                   row.names = F)
```

## WGCNA input

```{r WGCNA-input}

# Select unique treatment conditions

MVA_metadata <- metadata |> 
  filter(group == "experiment") |> 
  mutate(experiment = toupper(experiment),
         experiment_alphanum = gsub("[^[:alnum:]]", "", experiment)) |> 
  left_join(y = DESeq2Result |>
              distinct(experimentCondition, nSamplesDESeqDataSetFromMatrix) |> 
              tibble(), 
            by = join_by(experimentCondition)) |> 
  mutate(sample_id = paste0(params$system,"_",toupper(exposure_type),"_",experiment_alphanum,"_T", time_level, "_C", concentration_level),
         includedInMVA = TRUE) |> 
   dplyr::distinct(sample_id,
                  experiment,
                  exposure_type,
                  conc = concentration,
                  conc_level = concentration_level,
                  conc_unit,
                  time,
                  time_level,
                  time_unit, 
                  experimentCondition,
                  includedInMVA) |>
  drop_na()


MVA_input <- DESeq2Result |>  
  
  left_join(y = MVA_metadata |>  
              distinct(experimentCondition, sample_id), 
            by = join_by(experimentCondition)) |> 
  
  left_join(y = updateProbeResult |> 
              select(gene_symbol = "new_gene_symbol", 
                     entrez_id = "new_entrez_id") |> 
              distinct(),
            by = join_by(gene_symbol)) |> 
  select(sample_id, gene_symbol, entrez_id, "log2fc" = log2FoldChange, lfcSE, stat, pvalue, padj) |> 
  tibble() |> 
  mutate(gene_symbol = paste0("id_",gene_symbol),
         entrez_id = paste0("id_",entrez_id)) |> 
  left_join(y = customBasemean, join_by(gene_symbol)) |> 
  select(sample_id, gene_symbol, entrez_id, basemean, log2fc, pvalue, padj)

MVA_metadata <- MVA_metadata |> 
  select(-experimentCondition)


data.table::fwrite(x = MVA_input, 
                   file = file.path(outputPath, glue("{params$project}_MVA_input.txt")), 
                   sep = "\t", 
                   row.names = FALSE)


data.table::fwrite(x = MVA_metadata, 
                   file = file.path(outputPath, glue("{params$project}_MVA_metadata.txt")), 
                   sep = "\t", 
                   row.names = FALSE) 
```


## Save output
```{r Save}
#save.image(file = file.path(outputPath, glue("{params$project}_DAM_output_data_all.RDdata")))
save.image(paste0(outputPath,"/","_NEW4_",Sys.Date(),".RData"))

# save(contrastTable, concentrationtable, count, countdataCpm, countdataCpmSum, countdataRaw, countdataRawSum, customBasemean,
#      DESeq2Result, DESeq2Result_upload, metadata, probeManifest, relevanceFilterOutput, replicateCorrelationOutput_log2CPM,
#      updateProbeResult, includedSamplesDESeq2, MVA_input, MVA_metadata,
#      file = file.path(outputPath, glue("{params$project}_DAM_output_data.RDdata")))

rm(list = ls())
gc()
gc()
```
